pkg/elfreader/note.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2022 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package elfreader
+
+import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+const (
+	maxNoteSize        = 1 << 20 // in bytes
+	NoteTypeGNUBuildID = 3
+	NoteTypeGoBuildID  = 4
+)
+
+// ElfNote is the payload of a Note Section in an ELF file.
+type ElfNote struct {
+	Name string // Contents of the "name" field, omitting the trailing zero byte.
+	Desc []byte // Contents of the "desc" field.
+	Type uint32 // Contents of the "type" field.
+}
+
+// ParseNotes returns the notes from a SHT_NOTE section or PT_NOTE segment.
+func ParseNotes(reader io.Reader, alignment int, order binary.ByteOrder) ([]ElfNote, error) {
+	r := bufio.NewReader(reader)
+
+	// padding returns the number of bytes required to pad the given size to an
+	// alignment boundary.
+	padding := func(size int) int {
+		return ((size + (alignment - 1)) &^ (alignment - 1)) - size
```

Comment:
`&^` lol


actually... what does this do? 😄

Reply:
It seems that it's a logical NOT (+ AND)

> &^   bit clear (AND NOT)    integers

in the "Arithmetic operations" section from https://go.dev/ref/spec#Integer_operators